File: fcu.py
from sympy import symbols, Matrix, zeros

def main():
    # Matrix dimensions
    M, N = 3, 3

    # Setting up symbols
    Rsense, Rcol, Rrow = symbols('Rsense Rcol Rrow')
    G = [[symbols(f'g{i+1}{j+1}') for j in range(N)] for i in range(M)]
    V = [[symbols(f'V{i+1}{j+1}') for j in range(N)] for i in range(M)]
    Va = [[symbols(f'Va{i+1}{j+1}') for j in range(N)] for i in range(M)]
    Vin = [symbols(f'Vin{i+1}') for i in range(M)]

    # M, N = 3, 3
    # Rsense, Rcol, Rrow = 5, 2, 3
    # G = [[1/10, 1/15, 1/20], [1/25, 1/30, 1/35], [1/40, 1/45, 1/50]]
    # # G = [[0, 0, 0], [0, 0, 0], [1/10, 0, 0]]
    # V = [[((i+1)*(j+1)) for j in range(N)] for i in range(M)]
    # Va = [[((i+1)*(j+1))  for j in range(N)] for i in range(M)]
    # Vin = [5, 7, 9]

    # M, N = 2, 2
    # Rsense, Rcol, Rrow = 5, 2, 3
    # # G = [[1/10, 1/15, 1/20], [1/25, 1/30, 1/35], [1/40, 1/45, 1/50]]
    # G = [[1/11, 1/17], [1/29, 1/37]]
    # Vin = [7, 5]

    print("G:")
    for row in G:
        print(row)
    print()

    print("Vin:")
    for v in Vin:
        print(v)
    print()
    
    # ----- Step 1: Formulate Column Linear Systems -----
    A = []
    for j in range(N):
        Aj = []
        for row in range(M):
            row_elem = []
            for i in range(M):
                if i == row:
                    row_elem.append(-1)
                elif i > row:
                    row_elem.append((i - row) * G[i][j] * Rcol)
                else:
                    row_elem.append(0)
            Aj.append(row_elem)
        A.append(Matrix(Aj))
    
    for j in range(N):
        print("A", j, ":")
        for row in A[j].tolist():
            print(row)
        print()
    print()

    Vcol = []
    for j in range(N):
        Vcolj = []
        for i in range(M):
            Vcolj.append([V[i][j]])
        Vcol.append(Matrix(Vcolj))
    
    VAcol = []
    for j in range(N):
        VAcolj = []
        for i in range(M):
            VAcolj.append([Va[i][j]])
        VAcol.append(Matrix(VAcolj))
    
    J = Matrix([[(Rsense + i * Rcol)] for i in range(M).__reversed__()])

    print("J:")
    for j in J:
        print(j)
    print()
    
    K = []
    for j in range(N):
        K.append(Matrix([G[i][j] for i in range(M)]).T)

    print("K:")
    for j in range(N):
        print("K", j, ":")
        print(K[j])
        print()
    print()
    
    Iout = []
    for j in range(N):
        Ioutj = 0
        for i in range(M):
            Ioutj += G[i][j] * V[i][j]
        Iout.append(Ioutj)
    
    # -Aj * Vcolj = VAcolj - J * Ioutj
    for j in range(N):
        print(f"Column {j}:")
        for i in range(M):
            print((-A[j] * Vcol[j])[i], " = ", (VAcol[j] - J * Iout[j])[i])
        print()
    
    # (-Aj + J * Kj) * Vcolj = VAcolj
    for j in range(N):
        print(f"Column {j}:")
        for i in range(M):
            print(((-A[j] + J * K[j])*Vcol[j])[i], " = ", VAcol[j][i])
        print()

    for j in range(N):
        tmp = J * K[j]
        for row in tmp.tolist():
            print(row)
        print()
    
    # ----- Step 2: Merge Column Linear Systems -----
    COLmat_matrices = [(-A[j] + J * K[j]) for j in range(N)]
    COLmat = direct_sum(*COLmat_matrices)

    print("COLmat:")
    for row in COLmat.tolist():
        print(row)
    print()
    
    Gmat = direct_sum(*K)

    print("Gmat:")
    for row in Gmat.tolist():
        print(row)
    print()
    
    CVcol = Matrix.vstack(*Vcol)
    CVAcol = Matrix.vstack(*VAcol)

    # COLmat * CVcol = CVAcol
    for i in range(M*M):
        print((COLmat * CVcol)[i], " = ", CVAcol[i])
    print()

    # Gmat * CVcol = (Iout)^T
    for i in range(M):
        print((Gmat * CVcol)[i], " = ", Iout[i])
    print()

    # ----- Step 3: Formulate Row Linear Systems -----
    B = []
    for i in range(M):
        Bi = []
        for row in range(N):
            row_elem = []
            for j in range(N):
                if (row <= j):
                    row_elem.append(G[i][j] * (row + 1))
                else:
                    row_elem.append(G[i][j] * (j + 1))
            Bi.append(row_elem)
        B.append(Matrix(Bi) * Rrow)

    for i in range(M):
        print("B", i, ":")
        for row in B[i].tolist():
            print(row)
        print()
    print()

    Vrow = []
    for i in range(M):
        Vrowi = []
        for j in range(N):
            Vrowi.append([V[i][j]])
        Vrow.append(Matrix(Vrowi))

    VArow = []
    for i in range(M):
        VArowi = []
        for j in range(N):
            VArowi.append([Va[i][j]])
        VArow.append(Matrix(VArowi))

    Vrowin = [Matrix([Vin[i] for j in range(N)]) for i in range(M)]

    # Bi * Vrowi = Vrowini - VArowi
    for i in range(M):
        print(f"Row {i}:")
        for j in range(N):
            print((B[i] * Vrow[i])[j], " = ", (Vrowin[i] - VArow[i])[j])
        print()

    # ----- Step 4: Merge Row Linear Systems -----
    ROWmat = direct_sum(*B)

    print("ROWmat:")
    for row in ROWmat.tolist():
        print(row)
    print()

    CVrow = Matrix.vstack(*Vrow)
    CVArow = Matrix.vstack(*VArow)
    CVrowin = Matrix.vstack(*Vrowin)

    # CVrowin - ROWmat * CVrow = CVArow
    for j in range(N*N):
        print((CVrowin - ROWmat * CVrow)[j], " = ", CVArow[j])
    print()

    # ----- Step 5: Eliminate Internal Variables -----
    ROWmatAint = zeros(N*M, N*M)
    CVrowinA = zeros(N*M, 1)
    for j in range(N):
        for i in range(M):
            ROWmatAint[j*M + i, :] = ROWmat[i*N + j, :]
            CVrowinA[j*M + i, :] = CVrowin[i*N + j, :]
    ROWmatA = zeros(N*M, N*M)
    for j in range(N):
        for i in range(M):
            ROWmatA[:, j*M + i] = ROWmatAint[:, i*N + j]

    print("ROWmatA:")
    for row in ROWmatA.tolist():
        print(row)
    print()

    print("CVrowINA:")
    for v in CVrowinA:
        print(v)
    print()

    # CVrowINA - ROWmatA * CVcol = CVAcol
    for i in range(N*M):
        print((CVrowinA - ROWmatA * CVcol)[i], " = ", CVAcol[i])
    print()

    # tmp = COLmat + ROWmatA
    # tmp_inv = tmp.inv()

    # with open("tmp_inv.txt", "w") as f:
    #     for i in range(N*N):
    #         for j in range(M*M):
    #             print(f"({i} {j}):", file=f)
    #             print(tmp_inv[i, j], file=f)
    #             print(file=f)

    # for i in range(N*N):
    #     for j in range(M*M):
    #         print(f"({i} {j}):")
    #         print(tmp_inv[i, j])
    #         print()

    tmp = COLmat + ROWmatA
    print("COLmat + ROWmatA:")
    for row in tmp.tolist():
        print(row)
    print()

    print("(COLmat + ROWmatA)^-1:")
    for row in tmp.inv().tolist():
        print(row)
    print()

    NETmat = Gmat * (COLmat + ROWmatA).inv()

    print("NETmat:")
    for row in NETmat.tolist():
        print(row)
    print()
    
    # CVrowina - ROWmata * CVcol = CVAcol
    # for i in range(N*N):
    #     print((CVrowina - ROWmata * CVcol)[i], " = ", CVAcol[i])

    # (COLmat + ROWmatA) * CVcol = CVrowina
    # for i in range(N*N):
    #     print(((COLmat + ROWmata) * CVcol)[i], " = ", CVrowina[i])

    # (Iout)^T = NETmat * CVrowina
    # for i in range(M):
    #     print(Iout[i], " = ", (NETmat * CVrowina)[i])

    # ----- Step 6: Reduce Matrix Dimension -----
    NETmatC = zeros(N, M)
    for i in range(M):
        for j in range(N):
            NETmatC[:, i] += NETmat[:, j*M + i]

    print("NETmatC:")
    for row in NETmatC.tolist():
        print(row)
    print()

    Iout = NETmatC * Matrix(Vin)

    print("Iout:")
    for i in Iout:
        print(i.evalf())
    print()


def direct_sum(*matrices):
    # Calculate the total size of the resulting matrix
    total_rows = sum(m.rows for m in matrices)
    total_cols = sum(m.cols for m in matrices)
    
    # Create an empty matrix of the final size, filled with zeros
    result = zeros(total_rows, total_cols)
    
    # Position each matrix in its block-diagonal location
    row_offset = 0
    col_offset = 0
    for m in matrices:
        rows, cols = m.shape
        result[row_offset:row_offset + rows, col_offset:col_offset + cols] = m
        row_offset += rows
        col_offset += cols
    
    return result

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(e)
    
    input("Press Enter to exit...")

File: fcu.cpp
#include <iostream>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <vector>
#include <chrono>

// Based on:
// https://ieeexplore-ieee-org.tudelft.idm.oclc.org/document/9108292

Eigen::VectorXf solve_fcu(const Eigen::MatrixXf& G, const Eigen::VectorXf& Vin, const float Rcol, const float Rrow, const float Rsense, bool print = false) {
    // auto start_time = std::chrono::high_resolution_clock::now();

    int M = G.rows();
    int N = G.cols();

    // ----- Step 1: Formulate Column Linear Systems -----
    // A consists of N matrices, which are each MxM
    // std::vector<Eigen::MatrixXf> A;
    // for (int j = 0; j < N; j++) {
    //     Eigen::MatrixXf Aj = Eigen::MatrixXf::Zero(M, M);
    //     for (int row = 0; row < M; row++) {
    //         for (int i = 0; i < M; i++) {
    //             if (i == row) { Aj(row, i) = -1; }
    //             else if (i > row) { Aj(row, i) = (i - row) * G(i, j) * Rcol; }
    //         }
    //     }

    //     A.push_back(Aj);
    // }

    // if (print) {
    //     for (int j = 0; j < N; j++) {
    //         std::cout <<  "A" << j << std::endl << A[j] << std::endl << std::endl;
    //     }
    // }

    // // J is a column vector consisting of M elements
    // Eigen::VectorXf J(M);
    // for (int i = 0; i < M; i++) {
    //     J(i) = Rsense + (M - (i + 1)) * Rcol;
    // }

    // if (print) {
    //     std::cout << "J:\n" << J << std::endl << std::endl;
    // }

    // // K consists of N row vectors, each with M elements
    // std::vector<Eigen::MatrixXf> K;
    // for (int j = 0; j < N; j++) {
    //     Eigen::RowVectorXf Kj(M);
    //     for (int i = 0; i < M; i++) {
    //         Kj(i) = G(i, j);
    //     }
    //     K.push_back(Kj);
    // }

    // if (print) {
    //     for (int j = 0; j < N; j++) {
    //         std::cout <<  "K" << j << std::endl << K[j] << std::endl << std::endl;
    //     }
    // }

    // auto timestamp1 = std::chrono::high_resolution_clock::now();
    // auto execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp1 - start_time).count();
    // std::cout << "Setup time: " << execution_time << " (ms)" << std::endl;
    // timestamp1 = std::chrono::high_resolution_clock::now();

    // ----- Step 2: Merge Column Linear Systems -----
    // COLmat is the direct sum of (Aj + J * Kj), thus is a M*N x M*N matrix
    Eigen::MatrixXf COLmat = Eigen::MatrixXf::Zero(M*N, M*N);
    for (int j = 0; j < N; j++) {
        for (int row = 0; row < M; row++) {
            for (int i = 0; i < M; i++) {
                if (i == row) { COLmat(row + j*M, i + j*M) = 1 + (Rsense + (M - (row + 1)) * Rcol) * G(i, j); }
                else if (i > row) { COLmat(row + j*M, i + j*M) = -(i - row) * G(i, j) * Rcol + (Rsense + (M - (row + 1)) * Rcol) * G(i, j); }
                else { COLmat(row + j*M, i + j*M) = (Rsense + (M - (row + 1)) * Rcol) * G(i, j); }
            }
        }
        // Eigen::MatrixXf tmp = -A[j] + J * K[j];
        // COLmat.block(j*M, j*M, tmp.rows(), tmp.cols()) = tmp;
    }

    // auto timestamp2 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp2 - timestamp1).count();
    // std::cout << "COLmat time: " << execution_time << " (ms)" << std::endl;
    // timestamp2 = std::chrono::high_resolution_clock::now();

    if (print) {
        std::cout << "COLmat:\n" << COLmat << std::endl << std::endl;
    }

    // Gmat is the direct sum of Kj, thus is a N x M*N matrix
    Eigen::MatrixXf Gmat = Eigen::MatrixXf::Zero(N, M*N);
    for (int j = 0; j < N; j++) {
        for (int i = 0; i < M; i++) {
            Gmat(j, i + j*M) = G(i, j);
        }
    }

    // auto timestamp3 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp3 - timestamp2).count();
    // std::cout << "Gmat time: " << execution_time << " (ms)" << std::endl;
    // timestamp3 = std::chrono::high_resolution_clock::now();

    if (print) {
        std::cout << "Gmat:\n" << Gmat << std::endl << std::endl;
    }

    // ----- Step 3: Formulate Row Linear Systems -----
    // B consists of M matrices, which are each NxN
    // std::vector<Eigen::MatrixXf> B;
    // for (int i = 0; i < M; i++) {
    //     Eigen::MatrixXf Bi(N, N);
    //     for (int row = 0; row < N; row++) {
    //         for (int j = 0; j < N; j++) {
    //             if (row <= j) { Bi(row, j) = G(i, j) * (row + 1); }
    //             else { Bi(row, j) = G(i, j) * (j + 1); }
    //         }
    //     }

    //     B.push_back(Bi*Rrow);
    // }

    // if (print) {
    //     for (int i = 0; i < M; i++) {
    //         std::cout <<  "B" << i << std::endl << B[i] << std::endl << std::endl;
    //     }
    // }
    
    // ----- Step 4: Merge Row Linear Systems -----
    // ROWmat is the direct sum of all Bi matrices, thus is a N*M x N*M matrix
    Eigen::MatrixXf ROWmat = Eigen::MatrixXf::Zero(N*M, N*M);
    for (int i = 0; i < M; i++) {
        for (int row = 0; row < N; row++) {
            for (int j = 0; j < N; j++) {
                if (row <= j) { ROWmat(row + i*N, j + i*N) = G(i, j) * (row + 1); }
                else { ROWmat(row + i*N, j + i*N) = G(i, j) * (j + 1); }
            }
        }
        // ROWmat.block(i*N, i*N, B[i].rows(), B[i].cols()) = B[i];
    }
    ROWmat *= Rrow;

    // auto timestamp4 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp4 - timestamp3).count();
    // std::cout << "ROWmat time: " << execution_time << " (ms)" << std::endl;
    // timestamp4 = std::chrono::high_resolution_clock::now();

    if (print) {
        std::cout << "ROWmat:\n" << ROWmat << std::endl << std::endl;
    }

    // ----- Step 5: Eliminate Internal Variables -----
    // ROWmatA and CVrowINA are constructed by applying row swaps to ROWmat and CVrowIN respectively, thus also have the same shapes
    Eigen::MatrixXf ROWmatAint(N*M, N*M);
    for (int j = 0; j < N; j++) {
        for (int i = 0; i < M; i++) {
            ROWmatAint.row(j*M + i) = ROWmat.row(i*N + j);
        }
    }
    Eigen::MatrixXf ROWmatA(N*M, N*M);
    for (int j = 0; j < N; j++) {
        for (int i = 0; i < M; i++) {
            ROWmatA.col(j*M + i) = ROWmatAint.col(i*N + j);
        }
    }

    // auto timestamp5 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp5 - timestamp4).count();
    // std::cout << "ROWmatA time: " << execution_time << " (ms)" << std::endl;
    // timestamp5 = std::chrono::high_resolution_clock::now();

    if (print) {
        std::cout << "ROWmatA:\n" << ROWmatA << std::endl << std::endl;
    }

    // NETmat is constructe as such: Gmat * (COLmat + ROWmatA)^-1, thus is a N x N*M matrix
    // Eigen::MatrixXf NETmat = Gmat * (COLmat + ROWmatA).inverse();

    // Alternative:
    // Rewrite NEtmat = Gmat * (COLmat + ROWmatA)^-1 as (COLmat + ROWmatA)^T * NETmat^T = Gmat^T
    // Then use a solver to solve for NETmat^T
    // Finally transpose to get NETmat    
    Eigen::SparseMatrix<float> COLROWmat = (COLmat + ROWmatA).sparseView();
    Eigen::MatrixXf Gmat_t = Gmat.transpose();
    Eigen::SparseMatrix<float> Gmat_t_sparse = Gmat_t.sparseView();

    // auto timestamp6 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp6 - timestamp5).count();
    // std::cout << "Solver setup time: " << execution_time << " (ms)" << std::endl;
    // timestamp6 = std::chrono::high_resolution_clock::now();

    Eigen::BiCGSTAB<Eigen::SparseMatrix<float>> solver;

    solver.compute(COLROWmat.transpose());
    Eigen::SparseMatrix<float> NETmat = solver.solve(Gmat_t_sparse).transpose();

    // auto timestamp7 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp7 - timestamp6).count();
    // std::cout << "Solve time: " << execution_time << " (ms)" << std::endl;
    // timestamp7 = std::chrono::high_resolution_clock::now();

    if (print) {
        std::cout << "NETmat:\n" << NETmat << std::endl << std::endl;
    }

    // ----- Step 6: Reduce Matrix Dimensions -----
    // NETmatC is constructed by using column sums of NETmat such that CVrowINA can be reduced to Vin, thus is a NxM matrix
    Eigen::MatrixXf NETmatC = Eigen::MatrixXf::Zero(N, M);
    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            NETmatC.col(i) += NETmat.col(j*M + i);
        }
    }

    // auto timestamp8 = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(timestamp8 - timestamp7).count();
    // std::cout << "NETmatC time: " << execution_time << " (ms)" << std::endl;
    // timestamp8 = std::chrono::high_resolution_clock::now();

    if (print) {
        std::cout << "NETmatC:\n" << NETmatC << std::endl << std::endl;
    }

    Eigen::VectorXf Iout = NETmatC * Vin;

    // auto end_time = std::chrono::high_resolution_clock::now();
    // execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - timestamp8).count();
    // std::cout << "Iout time: " << execution_time << " (ms)" << std::endl;

    return Iout;
}

int main(int argc, char* argv[]) {
    long long total_time = 0;

    srand((unsigned int) time(0));

    int M = (argc >= 2) ? std::atoi(argv[1]) : 16;
    int N = (argc >= 3) ? std::atoi(argv[2]) : 16;

    int runs = (argc >= 4) ? std::atoi(argv[3]) : 1;

    bool print = false;
    if (std::atoi(argv[4]) == 1) { print = true; }

    for (int i = 0; i < runs; i++) {
        float Rmin = 100.;
        float Rmax = 1000.;
        float Vdd = 5.;

        float Rcol = 2.;
        float Rrow = 3.;
        float Rsense = 5.;

        if (print) {
            std::cout << "Rcol: " << Rcol << std::endl;
            std::cout << "Rrow: " << Rrow << std::endl;
            std::cout << "Rsense: " << Rsense << std::endl << std::endl;
        }

        // Construct G matrix with random values
        Eigen::MatrixXf R = Eigen::MatrixXf::Random(M, N);
        R = (R + Eigen::MatrixXf::Constant(M, N, 1.0)) * 0.5 * (Rmax - Rmin) + Eigen::MatrixXf::Constant(M, N, Rmin);
        R(0, 0) = 10.;
        R(0, 1) = 15.;
        R(0, 2) = 20.;
        R(1, 0) = 25.;
        R(1, 1) = 30.;
        R(1, 2) = 35.;
        R(2, 0) = 40.;
        R(2, 1) = 45.;
        R(2, 2) = 50.;

        if (print) {
            std::cout << "R:\n" << R << std::endl << std::endl;
        }

        Eigen::MatrixXf G = R.cwiseInverse();

        if (print) {
            std::cout << "G:\n" << G << std::endl << std::endl;
        }

        // Construct Vin with random values
        Eigen::VectorXf Vin = Eigen::VectorXf::Random(M);
        Vin = (Vin.array() > 0.5).select(Eigen::VectorXf::Constant(M, Vdd), Eigen::VectorXf::Zero(M));
        Vin(0) = 5;
        Vin(1) = 7;
        Vin(2) = 9;

        if (print) {
            std::cout << "Vin:\n" << Vin << std::endl << std::endl;
        }

        // Eigen::VectorXf Iout(N)

        auto start_time = std::chrono::high_resolution_clock::now();

        Eigen::VectorXf Iout = solve_fcu(G, Vin, Rcol, Rrow, Rsense, print);

        auto end_time = std::chrono::high_resolution_clock::now();

        if (print) {
            std::cout << Iout << std::endl << std::endl;
        }

        auto execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();
        
        std::cout << "Execution time: " << execution_time << " ms" << std::endl;

        total_time += execution_time;
    }

    if (runs > 1) {
        std::cout << "Average execution time: " << total_time/runs << " ms" << std::endl;
    }

    return 0;
}
